### ISSUE TITLE
uninitialized variable in loop gives incorrect error

### ISSUE THREAD

**Issue by squirrelyman** [rust-lang/rust#59015]
An uninitialized variable in a loop gives a confusing borrow checker error.

Example : 
```rust
loop {
        let mut input;
        std::io::stdin().read_line(&mut input).unwrap();
}
```

Expected:
```
error[E0381]: borrow of possibly uninitialized variable: input
```

Actual:
```
error[E0382]: borrow of moved value: `input`
 --> src/main.rs:5:36
  |
4 |         let mut input;
  |             --------- move occurs because `input` has type `std::string::String`, which does not implement the `Copy` trait
5 |         std::io::stdin().read_line(&mut input).unwrap();
  |                                    ^^^^^^^^^^ value borrowed here after move
6 | }
  | - value moved here, in previous iteration of loop
```

The move error is confusing because it implies the variable somehow got moved despite being declared right before it is used.

**Comment by steveklabnik** [comment]
Triage: no changes

**Comment by Aaron1011** [comment]
This now results the expected error message both with and without `#![feature(nll)]`.

**Comment by Enselic** [comment]
This was fixed in **nightly-2021-10-01**:

```console
$ cargo bisect-rustc --verbose --preserve --start=2021-09-01 --end=2022-04-01 --regress=success --script bash -- -c "cargo check 2>&1 | grep -i initialized"
checking the start range to find a passing nightly
installing nightly-2021-09-01
testing...
RESULT: nightly-2021-09-01, ===> Script returned error

[...]

searched toolchains nightly-2021-09-01 through nightly-2022-04-01

********************************************************************************
Regression in nightly-2021-10-01
********************************************************************************

[...]

ERROR: no CI builds available between 11491938f80988c7261a1179cf71a25c379c8783 and aa7aca3b954131720df725e70d12e902eb3be1de within last 167 days
```

Which leaves us with:

```console
$ git log --no-merges --oneline 11491938f80988c7261a1179cf71a25c379c8783..aa7aca3b954131720df725e70d12e902eb3be1de
0f4731273a2 Fix Windows LLVM issue.
762f81ddafa Update books
80b4718fd3a Update cargo
bdd34717b86 Remove Never variant from clean::Type enum
a0f4e783fcf Remove lazy_static dependency
d2613fb7a5b Improve help for recursion limit errors
830ecbd96ce Optimize is_sorted for Range and RangeInclusive
dd0b5f48158 Clarify that `CString::from_vec_unchecked` appends 0 byte.
cd0873b5027 Add unit assignment to MIR for `asm!()`
12c3f50a90b PassWrapper: handle function rename from upstream D36850
592d0892aa7 :arrow_up: rust-analyzer
d0e2b607de9 Fix test
87010206adc 2229: Consume IfLet expr
78013f296ac Don't anonymize bound region names during typeck
4d66986e090 Use larger span for adjustments on method calls
b226d1751d1 Fix generics where bounds order
29029c0bc23 Fix formatting
ff1ecc0ee9e Add tracking issue
92b57c04761 Updated for new const trait bounds syntax
3051bb9c811 Move tests to library/core/tests
aaf902bf8ff Add test for try operator with Option
a042705a7dd Constly impl TryV2 and Residual for Option
991b375d87c Add test for try operator on Result
88258c02a9c Constly impl TryV2 and FromResidual for Result
b85107ec711 Add tests for feature(const_identity_convert)
b82aaf49132 Constify identify conversions
6ff5b471ef7 Fix issue #72649: avoid spurious "previous iteration of loop" errors.
18fb97e4cc6 Remove ignore-tidy-undocumented-unsafe from core::slice::sort
ee3811671b8 Add test for union keyword highlighting
47c50560146 Fix highlighting for union keyword
```

Of which the most likely seems https://github.com/rust-lang/rust/commit/6ff5b471ef7, which added a bunch of tests that looks like the code in the original issue. So I don't see a need for more tests.

Closing as fixed.
